src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -210,14 +220,32 @@ private void TrackOpenedDocument(uint docCookie, string moniker)
 
         private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
+            // We may or may not be getting this notification from the foreground thread if another workspace
+            // is raising events on a background. Let's send it back to the UI thread since we can't talk
+            // to the RDT in the background thread. Since this is all asynchronous a bit more asynchrony is fine.
+            if (!_foregroundThreadAffinitization.IsForeground())
+            {
+                ScheduleTask(() => Registration_WorkspaceChanged(sender, e));
```

Comment:
I'm also trying to recall if this was an actual problem or I was just hardening against the theory, since I don't know what else is out there. The more direct problem was the ordering where a workspace would get registered/unregistered due to a rename and we saw the event before the RDT event had been raised.